discovery-provider/src/tasks/index_user_listening_history.py

```diff
@@ -0,0 +1,137 @@
+import logging
+import time
+from typing import DefaultDict
+import sqlalchemy as sa
+from src.models.models import UserListeningHistory
+from src.tasks.celery_app import celery
+from src.models import IndexingCheckpoints, Play
+from src.utils.helpers import UPDATE_INDEXING_CHECKPOINTS_QUERY
+
+logger = logging.getLogger(__name__)
+
+USER_LISTENING_HISTORY_TABLE_NAME = "user_listening_history"
+BATCH_SIZE = 100000 # index 100k plays at most at a time
+
+def sort_listening_history(deduped_history_dict, limit = 100):
+    # sorts listening history and places a limit
+    deduped_history = []
+    for track_id, timestamp in deduped_history_dict.items():
+        deduped_history.append({"track_id": track_id, "timestamp": timestamp})
+    deduped_history.sort(key = lambda listen: listen["timestamp"], reverse=True)
+    return deduped_history[:limit]
+
+def _index_user_listening_history(session):
+    # get the last updated id that counted towards user_listening_history
+    # use as lower bound
+    prev_id_checkpoint = (session.query(IndexingCheckpoints.last_checkpoint)
+        .filter(IndexingCheckpoints.tablename == USER_LISTENING_HISTORY_TABLE_NAME)
+    ).scalar()
+
+    if not prev_id_checkpoint:
+        prev_id_checkpoint = 0
+
+    # get new plays since the last checkpoint
+    new_plays = (
```

Comment:
super edge case, but if number of new plays > BATCH_SIZE within the interval we'll never catch up without some kind of pagination. with a batch size like 100k and this running every 5 seconds i don't see this being an issue however

Reply:
True... we have to make sure the batch size is larger than the new plays in a window. I think it's safe to let it run and catch up before moving queries over to it. If it falls behind too much, we should notice.